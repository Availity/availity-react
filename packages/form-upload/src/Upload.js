import React, { Suspense, useEffect, useCallback, useMemo, useState } from 'react';
import PropTypes from 'prop-types';
import UploadCore from '@availity/upload-core';
import { avFilesDeliveryApi } from '@availity/api-axios';
import { Input, InputGroup } from 'reactstrap';
import { FormGroup, Feedback } from '@availity/form';
import Icon from '@availity/icon';
import { useField, useFormikContext } from 'formik';
import classNames from 'classnames';

import FilePickerBtn from './FilePickerBtn';
import FileList from './FileList';
import '../styles.scss';

const Dropzone = React.lazy(() => import('react-dropzone'));

const dropzoneFallback = <div data-testid="dropzone-fallback">Loading...</div>;

const CLOUD_URL = '/cloud/web/appl/vault/upload/v1/resumable';
const CLOUD_URL_FUD = '/cloud/web/appl/file-upload-delivery/v1/batch/deliveries';

const Upload = ({
  allowedFileNameCharacters,
  allowedFileTypes,
  btnColor = 'light',
  btnText,
  bucketId,
  children,
  className,
  clientId,
  customerId,
  deliverFileOnSubmit = false,
  deliveryChannel,
  endpoint,
  onDeliverySuccess,
  onDeliveryError,
  disabled = false,
  feedbackClass,
  fileDeliveryMetadata,
  getDropRejectionMessage,
  max,
  maxSize,
  totalMaxSize,
  multiple = true,
  name,
  onFileRemove,
  onFilePreUpload = [],
  onFileUpload,
  showFileDrop = false,
  fallback = dropzoneFallback,
  isCloud,
}) => {
  const [field, metadata] = useField(name);
  const { errors, isSubmitting, isValidating, setFieldError, setFieldValue, setFieldTouched } = useFormikContext();
  const classes = classNames(
    className,
    metadata.touched ? 'is-touched' : 'is-untouched',
    metadata.touched && metadata.error && 'is-invalid'
  );
  const [totalSize, setTotalSize] = useState(0);

  const fieldValue = useMemo(() => (Array.isArray(field.value) ? field.value : []), [field]);

  const callFileDelivery = useCallback(
    async (upload) => {
      if (!Array.isArray(upload)) upload = [upload];
      const uploadResults = [];
      try {
        for (const u of upload) {
          const data = {
            deliveries: [
              {
                deliveryChannel,
                fileURI: isCloud ? u.s3References[0] : u.references[0],
                metadata: typeof fileDeliveryMetadata === 'function' ? fileDeliveryMetadata(u) : fileDeliveryMetadata,
              },
            ],
          };
          if (isCloud) {
            uploadResults.push(
              avFilesDeliveryApi.uploadFilesDelivery(data, {
                path: '',
                name: CLOUD_URL_FUD,
                clientId,
                customerId,
              })
            );
          } else {
            uploadResults.push(
              avFilesDeliveryApi.uploadFilesDelivery(data, {
                clientId,
                customerId,
              })
            );
          }
        }

        const responses = await Promise.all(uploadResults);
        if (typeof onDeliverySuccess === 'function') {
          onDeliverySuccess(responses);
        }
      } catch (error) {
        setFieldError(name, 'An error occurred while uploading files.');
        if (typeof onDeliveryError === 'function') {
          onDeliveryError(error);
        }
      }
    },
    [
      clientId,
      customerId,
      deliveryChannel,
      fileDeliveryMetadata,
      name,
      setFieldError,
      onDeliverySuccess,
      onDeliveryError,
      isCloud,
    ]
  );

  useEffect(() => {
    // eslint-disable-next-line unicorn/consistent-function-scoping
    async function checkValidFormAndCallFileDelivery() {
      if (Object.keys(errors).length === 0) {
        // deliver all on submit
        await callFileDelivery(fieldValue);
      }
    }
    if (
      !onFileUpload &&
      isSubmitting === true &&
      isValidating === false &&
      deliverFileOnSubmit &&
      deliveryChannel &&
      fileDeliveryMetadata
    ) {
      checkValidFormAndCallFileDelivery();
    }
  }, [
    callFileDelivery,
    deliverFileOnSubmit,
    deliveryChannel,
    errors,
    fieldValue,
    fileDeliveryMetadata,
    isSubmitting,
    isValidating,
    onFileUpload,
  ]);

  const validator = useCallback(
    (file) => {
      const isDuplicate = fieldValue.some((prev) => prev.file.name === file.name);
      if (isDuplicate) {
        return {
          code: 'duplicate-name',
          message: 'A file with this name already exists',
        };
      }

      return null;
    },
    [fieldValue]
  );

  const removeFile = (fileId, fileIndex) => {
    const newFiles = fieldValue.filter((file, index) => file.id !== fileId || index !== fileIndex);

    if (newFiles.length !== fieldValue.length) {
      setFieldValue(name, newFiles, true);
      const removedFile = fieldValue.find((file) => file.id === fileId);
      if (!removedFile.error && !removedFile.errorMessage) setTotalSize(totalSize - removedFile.file.size);
      if (onFileRemove) onFileRemove(newFiles, fileId);
    }
  };

  const setFiles = async (files) => {
    let selectedFiles = [];
    // FileList is not iterable
    // eslint-disable-next-line unicorn/no-for-loop
    for (let i = 0; i < files.length; i++) {
      selectedFiles[i] = files[i];
    }

    if (max && selectedFiles.length + fieldValue.length > max) {
      selectedFiles = selectedFiles.slice(0, Math.max(0, max - fieldValue.length));
    }

    const uploads = [...fieldValue];

    const options = {
      bucketId,
      customerId,
      clientId,
      onPreStart: onFilePreUpload,
      fileTypes: allowedFileTypes,
      maxSize,
      allowedFileNameCharacters,
    };

    if (isCloud) options.endpoint = CLOUD_URL;
    if (endpoint) options.endpoint = endpoint;

    let newTotalSize = 0;

    for await (const file of selectedFiles) {
      const upload = new UploadCore(file, options);
      await upload.generateId();

      const validationResult = validator(file);

      if (file.dropRejectionMessage) {
        upload.errorMessage = file.dropRejectionMessage;
      } else if (totalMaxSize && totalSize + newTotalSize + upload.file.size > totalMaxSize) {
        upload.errorMessage = 'Total documents size is too large';
      } else if (validationResult) {
        upload.errorMessage = validationResult.message;
      } else {
        upload.start();
        newTotalSize += upload.file.size;
      }

      if (onFileUpload) {
        onFileUpload(upload);
      } else if (deliveryChannel && fileDeliveryMetadata) {
        upload.onSuccess.push(() => {
          if (upload?.references?.[0]) {
            // allow form to revalidate when upload is complete
            setFieldTouched(name, true);
            // deliver upon upload complete, not form submit
            if (!deliverFileOnSubmit) {
              callFileDelivery(upload);
            }
          }
        });
      }

      uploads.push(upload);
    }

    setTotalSize(totalSize + newTotalSize);
    await setFieldValue(name, uploads, true);
  };

  const handleFileInputChange = async (event) => {
    event.stopPropagation();
    await setFiles(event.target.files);
  };

  const onDrop = async (acceptedFiles, fileRejections) => {
    const rejectedFilesToDrop = fileRejections.map(({ file, errors }) => {
      const dropRejectionMessage = getDropRejectionMessage
        ? getDropRejectionMessage(errors, file)
        : errors.map((error) => error.message).join(', ');

      file.dropRejectionMessage = dropRejectionMessage;
      return file;
    });
    await setFiles([...acceptedFiles, ...rejectedFilesToDrop]);
  };

  let fileAddArea;
  const text = btnText || (
    <>
      <Icon name="plus-circle" title="Add File Icon" />
      {fieldValue.length === 0 ? 'Add File' : 'Add Another File Attachment'}
    </>
  );

  if (!max || fieldValue.length < max) {
    fileAddArea = showFileDrop ? (
      <FormGroup for={name}>
        <Input name={name} style={{ display: 'none' }} />
        <InputGroup disabled={disabled} className={classes}>
          <Suspense fallback={fallback}>
            <Dropzone
              onDrop={onDrop}
              multiple={multiple}
              maxSize={maxSize}
              accept={allowedFileTypes}
              validator={validator}
            >
              {({ getRootProps, getInputProps, isDragActive }) => (
                <section>
                  <div
                    {...getRootProps({
                      className: isDragActive ? 'file-drop-active' : 'file-drop',
                    })}
                  >
                    <input data-testid="file-picker" {...getInputProps()} />
                    <p>
                      <strong>Drag and Drop</strong>
                    </p>
                    {text}
                  </div>
                </section>
              )}
            </Dropzone>
          </Suspense>
        </InputGroup>
        <Feedback className={classNames('d-block', feedbackClass)} name={name} />
      </FormGroup>
    ) : (
      <FilePickerBtn
        data-testid="file-picker"
        onChange={handleFileInputChange}
        color={fieldValue.length === 0 ? btnColor : 'link'}
        multiple={multiple}
        allowedFileTypes={allowedFileTypes}
        maxSize={maxSize}
        name={name}
        disabled={disabled}
      >
        {text}
      </FilePickerBtn>
    );
  }

  return (
    <>
      <FileList files={fieldValue} onRemoveFile={removeFile} data-testid="file-list">
        {children}
      </FileList>
      {fileAddArea}
    </>
  );
};

Upload.propTypes = {
  /** The ID of the bucket you want to upload to. */
  bucketId: PropTypes.string.isRequired,
  /** The ID obtained from APIConnect. Must be subscribed to the resumeable uploads API. */
  clientId: PropTypes.string.isRequired,
  /** The customer ID for the organization the user is uploading on behalf of. */
  customerId: PropTypes.string.isRequired,
  /** Identifies the field and matches the validation schema. */
  name: PropTypes.string.isRequired,
  /** Restrict the file name characters to a regex set. */
  allowedFileNameCharacters: PropTypes.string,
  /** Restrict the file types allowed to be uploaded to. eg: ['.jpeg', '.jpg']. */
  allowedFileTypes: PropTypes.arrayOf(PropTypes.string),
  /** The color of the button. Refer to the Reactstrap documentation to determine which colors are available. */
  btnColor: PropTypes.string,
  /** + Add File for initial file or + Add Another File Attachment if an attachment already have been selected. | The text that appears on the button. */
  btnText: PropTypes.node,
  children: PropTypes.func,
  className: PropTypes.string,
  /** This prop is used in tandem with deliveryChannel and fileDeliveryMetadata so that your files will only get delivered to fileDeliveryApi when the form is being submitted. When false, the files are delivered as the user adds them.  */
  deliverFileOnSubmit: PropTypes.bool,
  /** The name of the delivery channel that is unique to where you will deliver files via the avFileDeliveryApi. */
  deliveryChannel: PropTypes.string,
  /** Disable the file input. */
  disabled: PropTypes.bool,
  /** A custom fallback element to render while Dropzone is being imported from 'react-dropzone. Since Dropzone is only used when showFileDrop is true, it is imported using lazy loading and suspense to cut down on the bundle size the client needs to initially download. */
  fallback: PropTypes.node,
  feedbackClass: PropTypes.string,
  /** The metadata properties that have been configured for the delivery channel you are trying to reach with avFilesDeliveryApi. */
  fileDeliveryMetadata: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
  /** Override the default error message for files rejected when showFileDrop is true. */
  getDropRejectionMessage: PropTypes.func,
  /** Set if the component is being used in the cloud. This will override the URL being used. */
  isCloud: PropTypes.bool,
  /** Override the endpoint used for uploading the file(s) */
  endpoint: PropTypes.string,
  /** The maximum number of files allowed to be uploaded. 0 (or a falsey value) means unlimited. When the max number has been reached, the add button will disappear. */
  max: PropTypes.number,
  /** The maximum file size (in bytes) for a file to be uploaded. */
  maxSize: PropTypes.number,
  /** The total maximum combined file size (in bytes) for all the files to be uploaded. */
  totalMaxSize: PropTypes.number,
  /** Indicates that the user will be allowed to select multiple files when selecting files from the OS prompt. */
  multiple: PropTypes.bool,
  /** Callback to be executed when the delivery API call(s) have failed. It is called with the error that was thrown. */
  onDeliveryError: PropTypes.func,
  /** Callback to be executed when the delivery API call(s) have completed. It is called with an array of API responses (array of one if a single call was made). Note: a delivery can be REJECTED/FAILED/etc when the success callback is called - make sure to check the deliveryStatus for accurate handling. */
  onDeliverySuccess: PropTypes.func,
  onFilePreUpload: PropTypes.arrayOf(PropTypes.func),
  /** Callback called when file is removed. The callback is provided two arguments. 1. the updated files and 2. the id of the file that was removed. */
  onFileRemove: PropTypes.func,
  /** Callback to be executed when file is uploaded. The callback is provided the Upload instance from upload-core SDK. Use this callback to hook into the upload.onSuccess and upload.onError events and track which files have been uploaded and get references returned by the API if needed. */
  onFileUpload: PropTypes.func,
  /** Set as true to show a drag and drop file upload option instead of a button (file explorer still available on click). */
  showFileDrop: PropTypes.bool,
};

export default Upload;
