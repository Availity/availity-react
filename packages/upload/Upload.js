import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import UploadCore from '@availity/upload-core';
import Dropzone from 'react-dropzone';
import { InputGroup } from 'reactstrap';
import uuid from 'uuid/v4';
import FilePickerBtn from './FilePickerBtn';
import FileList from './FileList';
import './styles.scss';

const validationAttrs = ['min', 'max', 'required'];

class Upload extends Component {
  constructor(props) {
    super(props);

    this.state = {
      files: [],
    };
  }

  input = createRef();

  files = [];

  removeFile = fileId => {
    this.setState(({ files }) => {
      const newFiles = files.filter(file => file.id !== fileId);
      if (newFiles.length !== files.length) {
        this.files = newFiles;

        if (this.props.onFileRemove)
          this.props.onFileRemove(this.files, fileId);
        return {
          files: newFiles,
        };
      }
      return null;
    });
  };

  setFiles = files => {
    let selectedFiles = [];
    for (let i = 0; i < files.length; i++) {
      selectedFiles[i] = files[i];
    }
    if (
      this.props.max &&
      selectedFiles.length + this.state.files.length > this.props.max
    ) {
      selectedFiles = selectedFiles.slice(
        0,
        Math.max(0, this.props.max - this.state.files.length)
      );
    }
    this.files = this.files.concat(
      selectedFiles.map(file => {
        const upload = new UploadCore(file, {
          bucketId: this.props.bucketId,
          customerId: this.props.customerId,
          clientId: this.props.clientId,
          fileTypes: this.props.allowedFileTypes,
          maxSize: this.props.maxSize,
          allowedFileNameCharacters: this.props.allowedFileNameCharacters,
        });
        upload.id = `${upload.id}-${uuid()}`;
        if (file.dropRejectionMessage) {
          upload.errorMessage = file.dropRejectionMessage;
        } else {
          upload.start();
        }
        if (this.props.onFileUpload) this.props.onFileUpload(upload);
        return upload;
      })
    );
    this.setState({ files: this.files });
  };

  handleFileInputChange = event => {
    this.setFiles(event.target.files);
  };

  onDrop = (acceptedFiles, fileRejections) => {
    const rejectedFilesToDrop = fileRejections.map(({ file, errors }) => {
      const dropRejectionMessage = this.props.getDropRejectionMessage
        ? this.props.getDropRejectionMessage(errors, file)
        : errors.map(error => error.message).join(', ');

      file.dropRejectionMessage = dropRejectionMessage;
      return file;
    });
    this.setFiles([...acceptedFiles, ...rejectedFilesToDrop]);
  };

  reset = () => {
    this.files = [];
    this.setState({ files: [] });
  };

  componentDidMount() {
    if (this.context.FormCtrl && this.props.name) {
      this.updateValidations();
    }
  }

  updateValidations(props = this.props) {
    this.validations = { ...props.validate };

    Object.keys(props)
      .filter(val => validationAttrs.indexOf(val) > -1)
      .forEach(attr => {
        if (props[attr]) {
          this.validations[attr] = this.validations[attr] || {
            value: props[attr],
          };
        } else {
          delete this.validations[attr];
        }
      });

    this.context.FormCtrl.register(this);
    this.validate();
  }

  validate() {
    if (this.context.FormCtrl && this.props.name) {
      this.context.FormCtrl.validate(this.props.name);
    }
  }

  componentWillUnmount() {
    if (this.context.FormCtrl && this.props.name)
      this.context.FormCtrl.unregister(this);
  }

  getValue() {
    if (!this.files) return [];
    return this.files;
  }

  render() {
    const {
      btnText,
      max,
      multiple,
      allowedFileTypes,
      maxSize,
      children,
      showFileDrop,
      disabled,
    } = this.props;
    const { files } = this.state;

    let fileAddArea;
    const text = btnText || (
      <>
        <i className="icon icon-plus-circle" title="Add File Icon" />
        {files.length === 0 ? 'Add File' : 'Add Another File Attachment'}
      </>
    );

    if (!max || files.length < max) {
      if (showFileDrop) {
        fileAddArea = (
          <div>
            <InputGroup disabled={disabled}>
              <Dropzone
                onDrop={this.onDrop}
                multiple={multiple}
                maxSize={maxSize}
                className="file-drop"
                activeClassName="file-drop-active"
                accept={allowedFileTypes}
              >
                {({ getRootProps, getInputProps }) => (
                  <section>
                    <div {...getRootProps()}>
                      <input data-testid="file-picker" {...getInputProps()} />
                      <p>
                        <strong>Drag and Drop</strong>
                      </p>
                      {text}
                    </div>
                  </section>
                )}
              </Dropzone>
            </InputGroup>
          </div>
        );
      } else {
        fileAddArea = (
          <FilePickerBtn
            data-testid="file-picker"
            onChange={this.handleFileInputChange}
            color={files.length === 0 ? 'light' : 'link'}
            multiple={multiple}
            allowedFileTypes={allowedFileTypes}
            maxSize={maxSize}
            disabled={disabled}
          >
            {text}
          </FilePickerBtn>
        );
      }
    }

    return (
      <>
        <FileList files={files} onRemoveFile={this.removeFile}>
          {children}
        </FileList>
        {fileAddArea}
      </>
    );
  }
}

Upload.propTypes = {
  btnText: PropTypes.node,
  bucketId: PropTypes.string.isRequired,
  customerId: PropTypes.string.isRequired,
  clientId: PropTypes.string.isRequired,
  allowedFileNameCharacters: PropTypes.string,
  allowedFileTypes: PropTypes.arrayOf(PropTypes.string),
  onFileUpload: PropTypes.func,
  onFileRemove: PropTypes.func,
  maxSize: PropTypes.number,
  max: PropTypes.number,
  multiple: PropTypes.bool,
  children: PropTypes.func,
  name: PropTypes.string,
  showFileDrop: PropTypes.bool,
  getDropRejectionMessage: PropTypes.func,
  disabled: PropTypes.bool,
};

Upload.defaultProps = {
  multiple: true,
  showFileDrop: false,
  disabled: false,
};

export default Upload;
