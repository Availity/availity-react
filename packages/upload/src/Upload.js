/* eslint-disable react/no-unused-class-component-methods */
import React, { Component, createRef } from 'react';
import PropTypes from 'prop-types';
import UploadCore from '@availity/upload-core';
import Dropzone from 'react-dropzone';
import { InputGroup } from 'reactstrap';

import FilePickerBtn from './FilePickerBtn';
import FileList from './FileList';
import '../styles.scss';

const validationAttrs = ['min', 'max', 'required'];
const CLOUD_URL = '/cloud/web/appl/vault/upload/v1/resumable';

class Upload extends Component {
  constructor(props) {
    super(props);

    this.state = {
      files: [],
    };
  }

  input = createRef();

  files = [];

  removeFile = (fileId) => {
    this.setState(({ files }) => {
      const newFiles = files.filter((file) => file.id !== fileId);
      if (newFiles.length !== files.length) {
        this.files = newFiles;

        if (this.props.onFileRemove) this.props.onFileRemove(this.files, fileId);
        return {
          files: newFiles,
        };
      }
      return null;
    });
  };

  setFiles = async (files) => {
    let selectedFiles = [];
    // FileList is not iterable
    // eslint-disable-next-line unicorn/no-for-loop
    for (let i = 0; i < files.length; i++) {
      selectedFiles[i] = files[i];
    }
    if (this.props.max && selectedFiles.length + this.state.files.length > this.props.max) {
      selectedFiles = selectedFiles.slice(0, Math.max(0, this.props.max - this.state.files.length));
    }
    this.files = await Promise.all(
      // eslint-disable-next-line unicorn/prefer-spread
      this.files.concat(
        selectedFiles.map(async (file) => {
          const headers =
            typeof this.props.customHeaders === 'function' ? this.props.customHeaders(file) : this.props.customHeaders;
          const options = {
            bucketId: this.props.bucketId,
            customerId: this.props.customerId,
            clientId: this.props.clientId,
            fileTypes: this.props.allowedFileTypes,
            maxSize: this.props.maxSize,
            onPreStart: this.props.onFilePreUpload || [],
            allowedFileNameCharacters: this.props.allowedFileNameCharacters,
            headers,
          };

          if (this.props.endpoint) options.endpoint = this.props.endpoint;
          if (this.props.isCloud) options.endpoint = CLOUD_URL;

          const upload = new UploadCore(file, options);
          await upload.generateId();

          if (file.dropRejectionMessage) {
            upload.errorMessage = file.dropRejectionMessage;
          } else {
            upload.start();
          }

          if (this.props.onFileUpload) this.props.onFileUpload(upload);

          return upload;
        })
      )
    );
    this.setState({ files: this.files });
  };

  handleFileInputChange = async (event) => {
    await this.setFiles(event.target.files);
  };

  onDrop = async (acceptedFiles, fileRejections) => {
    const rejectedFilesToDrop = fileRejections.map(({ file, errors }) => {
      const dropRejectionMessage = this.props.getDropRejectionMessage
        ? this.props.getDropRejectionMessage(errors, file)
        : errors.map((error) => error.message).join(', ');

      file.dropRejectionMessage = dropRejectionMessage;
      return file;
    });
    await this.setFiles([...acceptedFiles, ...rejectedFilesToDrop]);
  };

  reset = () => {
    this.files = [];
    this.setState({ files: [] });
  };

  componentDidMount() {
    if (this.context.FormCtrl && this.props.name) {
      this.updateValidations();
    }
  }

  updateValidations(props = this.props) {
    this.validations = { ...props.validate };

    Object.keys(props)
      .filter((val) => validationAttrs.indexOf(val) > -1)
      // eslint-disable-next-line unicorn/no-array-for-each
      .forEach((attr) => {
        if (props[attr]) {
          this.validations[attr] = this.validations[attr] || {
            value: props[attr],
          };
        } else {
          delete this.validations[attr];
        }
      });

    this.context.FormCtrl.register(this);
    this.validate();
  }

  validate() {
    if (this.context.FormCtrl && this.props.name) {
      this.context.FormCtrl.validate(this.props.name);
    }
  }

  componentWillUnmount() {
    if (this.context.FormCtrl && this.props.name) this.context.FormCtrl.unregister(this);
  }

  getValue() {
    if (!this.files) return [];
    return this.files;
  }

  render() {
    const {
      btnColor,
      btnText,
      max,
      multiple,
      allowedFileTypes,
      maxSize,
      children,
      showFileDrop,
      disabled,
      onPasswordSubmit,
      passwordModalZIndex,
    } = this.props;
    const { files } = this.state;

    let fileAddArea;
    const text = btnText || (
      <>
        <i className="icon icon-plus-circle" title="Add File Icon" />
        {files.length === 0 ? 'Add File' : 'Add Another File Attachment'}
      </>
    );

    if (!max || files.length < max) {
      fileAddArea = showFileDrop ? (
        <div>
          <InputGroup disabled={disabled}>
            <Dropzone onDrop={this.onDrop} multiple={multiple} maxSize={maxSize} accept={allowedFileTypes}>
              {({ getRootProps, getInputProps, isDragActive }) => (
                <section>
                  <div
                    {...getRootProps({
                      className: isDragActive ? 'file-drop-active' : 'file-drop',
                    })}
                  >
                    <input data-testid="file-picker" {...getInputProps()} />
                    <p>
                      <strong>Drag and Drop</strong>
                    </p>
                    {text}
                  </div>
                </section>
              )}
            </Dropzone>
          </InputGroup>
        </div>
      ) : (
        <FilePickerBtn
          data-testid="file-picker"
          onChange={this.handleFileInputChange}
          color={files.length === 0 ? btnColor : 'link'}
          multiple={multiple}
          allowedFileTypes={allowedFileTypes}
          maxSize={maxSize}
          disabled={disabled}
        >
          {text}
        </FilePickerBtn>
      );
    }

    return (
      <>
        <FileList
          files={files}
          onRemoveFile={this.removeFile}
          onPasswordSubmit={onPasswordSubmit}
          passwordModalZIndex={passwordModalZIndex}
        >
          {children}
        </FileList>
        {fileAddArea}
      </>
    );
  }
}

Upload.propTypes = {
  /** The ID of the bucket you want to upload to. */
  bucketId: PropTypes.string.isRequired,
  /** The customer ID for the organization the user is uploading on behalf of. */
  customerId: PropTypes.string.isRequired,
  /** The ID obtained from APIConnect. Must be subscribed to the resumeable uploads API. */
  clientId: PropTypes.string.isRequired,
  /** The color of the button. Refer to the Reactstrap documentation to determine which colors
   * are available. **Default:** `light`. */
  btnColor: PropTypes.string,
  /** `+ Add File` for initial file or `+ Add Another File Attachment` if an attachment
   * already have been selected. | The text that appears on the button. */
  btnText: PropTypes.node,
  /** Restrict the file name characters to a regex set. */
  allowedFileNameCharacters: PropTypes.string,
  /** Restrict the file types allowed to be uploaded to. eg: `['.jpeg', '.jpg']`. */
  allowedFileTypes: PropTypes.arrayOf(PropTypes.string),
  /** Callback to be executed when file is uploaded. The callback is provided the `Upload`
   * instance from upload-core SDK. Use this callback to hook into the `upload.onSuccess`
   * and `upload.onError` events and track which files have been uploaded and get references
   * returned by the API if needed. See [example callback function above](###Callback-Function-Example) */
  onFileUpload: PropTypes.func,
  /** An array of callbacks called before the file is uploaded. Each function should return a true
   * or false. If one of these is false, then it will not upload. Useful for scanning files for
   * corruption before uploading to the vault. To ensure that the file is not uploaded and return
   * false, call setError on the UploadCore object to set some error message via calling
   * upload.setError('rejected', somemessage) method for updating the UI. Make sure this is done
   * relatively synchronously. */
  onFilePreUpload: PropTypes.arrayOf(PropTypes.func),
  /** Callback called when file is removed. The callback is provided two arguments. 1. the updated
   * files and 2. the id of the file that was removed */
  onFileRemove: PropTypes.func,
  /** The maximum file size (in bytes) for a file to be uploaded. */
  maxSize: PropTypes.number,
  /** The maximum number of files allowed to be uploaded. `0` (or a falsey value) means unlimited.
   * When the max number has been reached, the add button will disappear. */
  max: PropTypes.number,
  /** Indicates that the user will be allowed to select multiple files when selecting files from
   * the OS prompt. **Default:** `true`. */
  multiple: PropTypes.bool,
  /** Children can be a react child or render prop. */
  children: PropTypes.func,
  /** Identifies the field and matches the validation schema. */
  name: PropTypes.string,
  /** Set as true to show a drag and drop file upload option instead of a button (file explorer
   * still available on click). */
  showFileDrop: PropTypes.bool,
  /** Override the default error message for files rejected when `showFileDrop` is `true`. */
  getDropRejectionMessage: PropTypes.func,
  /** Disable the file input **Default:** `false`. */
  disabled: PropTypes.bool,
  /** When a user uploads an encrypted file, they are prompted to input a password. This function
   * is called when the password form is submitted. By default, the event bubbles and will submit
   * a form if the upload component is a child element of that form. Useful for adding
   * event.stopPropagation() if this behavior is not desired. */
  onPasswordSubmit: PropTypes.func,
  /** Override the default z-index for the password prompt modal. Useful for squashing IE11 bugs
   * by setting to auto if your upload component is already inside another modal. */
  passwordModalZIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Set if the component is being used in the cloud. This will override the URL being used. */
  isCloud: PropTypes.bool,
  /** Override the endpoint used for uploading the file(s) */
  endpoint: PropTypes.string,
  /** Set custom headers on the upload request */
  customHeaders: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
};

Upload.defaultProps = {
  btnColor: 'light',
  multiple: true,
  showFileDrop: false,
  disabled: false,
};

export default Upload;
