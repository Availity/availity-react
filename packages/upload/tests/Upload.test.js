import React from 'react';
import { render, fireEvent, cleanup, wait } from '@testing-library/react';
import Upload from '..';

afterEach(cleanup);

describe('Upload', () => {
  test('should render', () => {
    const { container } = render(
      <Upload clientId="a" bucketId="b" customerId="c" />
    );

    expect(container).toMatchSnapshot();
  });

  test('adding a file', () => {
    const { getByTestId } = render(
      <Upload clientId="a" bucketId="b" customerId="c" showFileDrop />
    );

    const file = new Buffer.from('hello world'.split('')); // eslint-disable-line new-cap
    file.name = 'fileName.png';
    const fileEvent = { target: { files: [file] } };

    const inputNode = getByTestId('file-picker');

    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
  });

  test('removing a file', () => {
    const { getByTestId, queryByTestId } = render(
      <Upload clientId="a" bucketId="b" customerId="c" />
    );

    // Create a new file
    const file = new Buffer.from('hello world'.split('')); // eslint-disable-line new-cap
    file.name = 'fileName.png';

    const fileEvent = { target: { files: [file] } };

    const inputNode = getByTestId('file-picker');

    // Simulate the upload to the Components
    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);

    const filerow = getByTestId('remove-file-btn');

    fireEvent.click(filerow);

    expect(queryByTestId('remove-file-btn')).toBeNull();
  });

  test('calls onFileRemove callback', () => {
    const mockFunc = jest.fn();

    const { getByTestId } = render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        onFileRemove={mockFunc}
      />
    );

    const inputNode = getByTestId('file-picker');

    const file = new Buffer.from('hello world'.split('')); // eslint-disable-line new-cap
    file.name = 'fileName.png';
    const fileEvent = { target: { files: [file] } };

    fireEvent.change(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);

    const filerow = getByTestId('remove-file-btn');

    fireEvent.click(filerow);

    expect(mockFunc).toHaveBeenCalled();
  });

  test('adds file via dropzone', () => {
    const { getByTestId } = render(
      <Upload clientId="a" bucketId="b" customerId="c" showFileDrop />
    );
    const file = new Buffer.from('hello world'.split('')); // eslint-disable-line new-cap
    file.name = 'fileName.png';

    const inputNode = getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
  });

  test('uses default drop rejection message', async () => {
    const { getByTestId, getByText } = render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        showFileDrop
        maxSize={10}
      />
    );
    const file = new Buffer.from('hello world'.split('')); // eslint-disable-line new-cap
    file.name = 'fileName.png';
    file.size = 11;

    const inputNode = getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
    await wait(() => {
      expect(getByText('File is larger than 10 bytes')).toBeDefined();
    });
  });

  test('uses custom drop rejection message', async () => {
    const getDropRejectionMessage = errors => {
      let msg = '';
      errors.forEach(error => {
        if (error.code === 'file-too-large') {
          msg += 'my custom error message';
        } else {
          msg += 'this file is no good';
        }
      });
      return msg;
    };

    const { getByTestId, getByText } = render(
      <Upload
        clientId="a"
        bucketId="b"
        customerId="c"
        showFileDrop
        maxSize={10}
        getDropRejectionMessage={getDropRejectionMessage}
      />
    );
    const file = new Buffer.from('hello world'.split('')); // eslint-disable-line new-cap
    file.name = 'fileName.png';
    file.size = 11;

    const inputNode = getByTestId('file-picker');
    const fileEvent = { target: { files: [file] } };

    fireEvent.drop(inputNode, fileEvent);

    expect(inputNode.files.length).toBe(1);
    await wait(() => {
      expect(getByText('my custom error message')).toBeDefined();
    });
  });
});
