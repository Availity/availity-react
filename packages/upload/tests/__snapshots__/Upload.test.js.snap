// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload should render 1`] = `
<div>
  <div
    class="d-none"
  >
    <input
      class="form-control-file"
      data-testid="file-picker"
      id="filePicker-1"
      multiple=""
      type="file"
      value=""
    />
  </div>
  <button
    class="btn btn-light"
    type="button"
  >
    <i
      class="icon icon-plus-circle"
      title="Add File Icon"
    />
    Add File
  </button>
</div>
`;
